atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -393,42 +393,70 @@ public void multiPut(Map<TableReference, ? extends Map<Cell, byte[]>> valuesByTa
     }
 
     @Override
     public void putUnlessExists(TableReference tableRef, Map<Cell, byte[]> values) throws KeyAlreadyExistsException {
         putInternal(
                 tableRef,
                 KeyValueServices.toConstantTimestampValues(values.entrySet(), AtlasDbConstants.TRANSACTION_TS),
-                true);
+                OverwriteBehaviour.DO_NOT_OVERWRITE);
+    }
+
+    @Override
+    public void setOnce(TableReference tableRef, Map<Cell, byte[]> values) {
+        putInternal(
+                tableRef,
+                KeyValueServices.toConstantTimestampValues(values.entrySet(), AtlasDbConstants.TRANSACTION_TS),
+                OverwriteBehaviour.OVERWRITE);
     }
 
     private void putInternal(
-            TableReference tableRef, Collection<Map.Entry<Cell, Value>> values, boolean doNotOverwriteWithSameValue) {
+            TableReference tableRef, Collection<Map.Entry<Cell, Value>> values, OverwriteBehaviour overwriteBehaviour) {
         Table table = getTableMap(tableRef);
         List<Cell> knownSuccessfullyCommittedKeys = new ArrayList<>();
         for (Map.Entry<Cell, Value> entry : values) {
             byte[] contents = entry.getValue().getContents();
             long timestamp = entry.getValue().getTimestamp();
 
             Key key = getKey(table, entry.getKey(), timestamp);
-            byte[] oldContents = putIfAbsent(table, key, contents);
-            if (oldContents != null && (doNotOverwriteWithSameValue || !Arrays.equals(oldContents, contents))) {
-                throw new KeyAlreadyExistsException(
-                        "We already have a value for this timestamp",
-                        ImmutableList.of(entry.getKey()),
-                        knownSuccessfullyCommittedKeys);
+            if (overwriteBehaviour != OverwriteBehaviour.OVERWRITE) {
+                byte[] oldContents = putIfAbsent(table, key, contents);
+                if (shouldThrow(overwriteBehaviour, contents, oldContents)) {
+                    throw new KeyAlreadyExistsException(
+                            "We already have a value for this timestamp",
+                            ImmutableList.of(entry.getKey()),
+                            knownSuccessfullyCommittedKeys);
+                }
+                knownSuccessfullyCommittedKeys.add(entry.getKey());
+            } else {
+                table.entries.put(key, copyOf(contents));
             }
-            knownSuccessfullyCommittedKeys.add(entry.getKey());
         }
     }
 
+    private boolean shouldThrow(OverwriteBehaviour overwriteBehaviour, byte[] contents, byte[] oldContents) {
+        return oldContents != null
+                && (overwriteBehaviour == OverwriteBehaviour.DO_NOT_OVERWRITE
+                        || (overwriteBehaviour == OverwriteBehaviour.OVERWRITE_SAME_VALUE
+                                && !Arrays.equals(oldContents, contents)));
```

Comment:
For readability, could we separate this out into terms? I almost prefer the less concise, if more direct approach:

```
if empty return false
switch (OB) {
  case DO_NOT_OVERWRITE: return true;
  case OVERWRITE_SAME_VALUE: return !Arrays.equals(...);
  case OVERWRITE: return false;
  default: throw new IllegalStateException();
}
```